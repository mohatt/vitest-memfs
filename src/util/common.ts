import { vi, expect } from 'vitest'

/**
 * Imports the real `fs/promises` module, bypassing Vitest mocks.
 *
 * A direct import works in most cases, but if `fs` is mocked
 * in a Vitest setup file it will return the mock instead.
 */
export async function importActualFS() {
  return vi.importActual<typeof import('fs/promises')>('fs/promises')
}

export function isPlainObject(value: unknown): value is Record<string, any> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

type MatchersObject = Parameters<(typeof expect)['extend']>[0]

/**
 * Creates and returns a matcher function.
 * We could wrap the matcher function in the future to warn on un-awaited promises.
 */
export function createMatcher<T extends keyof MatchersObject>(matcher: T, fn: MatchersObject[T]) {
  Object.defineProperty(fn, 'name', { value: matcher })
  return fn
}
