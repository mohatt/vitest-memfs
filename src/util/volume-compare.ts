import { createHash } from 'node:crypto'
import { isText } from 'istextorbinary'
import type { VolumeMap, VolumeMapEntry } from './volume.js'

export type VolumeCompareListMatch = 'exact' | 'ignore-extra' | 'ignore-missing'
export type VolumeCompareReportType = 'first' | 'all'

export interface VolumeCompareOptions {
  listMatch?: VolumeCompareListMatch
  report?: VolumeCompareReportType
}

export interface VolumeCompareResult {
  pass: boolean
  message: () => string
  actual?: unknown
  expected?: unknown
}

/**
 * Compare two volume maps.
 *
 * @param received map from volume under test
 * @param expected map from reference (snapshot or other volume)
 * @param options extra options
 */
export function compareVolumeMaps(
  received: VolumeMap,
  expected: VolumeMap,
  options?: VolumeCompareOptions,
): VolumeCompareResult {
  if (options?.report === 'all') {
    return compareVolumeMapsFull(received, expected, options)
  }

  const listMatch = options?.listMatch
  // make sorted arrays for error reporting and better diffing
  const actualFiles = Object.keys(received).sort()
  const expectedFiles = Object.keys(expected).sort()

  let listMismatchResult: {
    reason: string
    actual: string[]
    expected: string[]
  } | null = null

  switch (listMatch) {
    case 'ignore-extra': {
      const missing = expectedFiles.filter((f) => !(f in received))
      if (missing.length > 0) {
        listMismatchResult = {
          reason: `volume is missing ${missing.length} expected file${missing.length > 1 ? 's' : ''}`,
          actual: actualFiles.filter((f) => f in expected),
          expected: expectedFiles,
        }
      }
      break
    }

    case 'ignore-missing': {
      const extra = actualFiles.filter((f) => !(f in expected))
      if (extra.length > 0) {
        listMismatchResult = {
          reason: `volume has ${extra.length} unexpected file${extra.length > 1 ? 's' : ''}`,
          actual: actualFiles,
          expected: expectedFiles.filter((f) => f in received),
        }
      }
      break
    }

    case 'exact':
    default: {
      if (
        actualFiles.length !== expectedFiles.length ||
        !actualFiles.every((f, i) => f === expectedFiles[i])
      ) {
        listMismatchResult = {
          reason: 'directory structure didnâ€™t match',
          actual: actualFiles,
          expected: expectedFiles,
        }
      }
    }
  }

  if (listMismatchResult) {
    return {
      pass: false,
      message: () => listMismatchResult.reason,
      actual: listMismatchResult.actual,
      expected: listMismatchResult.expected,
    }
  }

  // compare contents
  const filesToCheck = listMatch === 'ignore-missing' ? actualFiles : expectedFiles
  for (const file of filesToCheck) {
    const exp = expected[file]
    const act = received[file]
    if (exp.type !== act.type) {
      return {
        pass: false,
        message: () => `path type mismatch at \`${file}\``,
        actual: makeEntryPreview(file, act),
        expected: makeEntryPreview(file, exp),
      }
    }
    if (exp.type === 'file') {
      const expBuff = exp.data
      const actBuff = (act as typeof exp).data
      if (!expBuff.equals(actBuff)) {
        return {
          pass: false,
          message: () => `mismatch in file \`${file}\``,
          actual: makeEntryPreview(file, act),
          expected: makeEntryPreview(file, exp),
        }
      }
    } else if (exp.type === 'symlink') {
      const expTarget = exp.target
      const actTarget = (act as typeof exp).target

      if (expTarget !== actTarget) {
        return {
          pass: false,
          message: () => `symlink target mismatch at \`${file}\``,
          actual: makeEntryPreview(file, act),
          expected: makeEntryPreview(file, exp),
        }
      }
    }
  }

  // everything is matched at this point
  return {
    pass: true,
    message: () => 'volumes matched',
  }
}

export function compareVolumeMapsFull(
  received: VolumeMap,
  expected: VolumeMap,
  options?: Omit<VolumeCompareOptions, 'report'>,
) {
  const listMatch = options?.listMatch
  const actualDiff: Record<string, unknown> = {}
  const expectedDiff: Record<string, unknown> = {}
  let missingCount = 0
  let extraCount = 0
  let diffCount = 0

  function addDiff(f: string, exp?: VolumeMapEntry, act?: VolumeMapEntry) {
    if (act) actualDiff[f] = makeEntryPreview(f, act)
    if (exp) expectedDiff[f] = makeEntryPreview(f, exp)
  }

  function addMatch(f: string) {
    actualDiff[f] = {}
    expectedDiff[f] = {}
  }

  const filesToCheck =
    listMatch === 'ignore-extra'
      ? expected
      : listMatch === 'ignore-missing'
        ? received
        : { ...received, ...expected }
  for (const f in filesToCheck) {
    const exp = expected[f]
    const act = received[f]

    if (exp && act) {
      if (exp.type !== act.type) {
        addDiff(f, exp, act)
        diffCount++
      } else if (exp.type === 'file') {
        if (!exp.data.equals((act as typeof exp).data)) {
          addDiff(f, exp, act)
          diffCount++
        } else {
          addMatch(f)
        }
      } else if (exp.type === 'symlink') {
        if (exp.target !== (act as typeof exp).target) {
          addDiff(f, exp, act)
          diffCount++
        } else {
          addMatch(f)
        }
      } else {
        addMatch(f)
      }
    } else if (exp && !act && listMatch !== 'ignore-missing') {
      addDiff(f, exp)
      missingCount++
    } else if (act && !exp && listMatch !== 'ignore-extra') {
      addDiff(f, null, act)
      extraCount++
    }
  }

  const total = missingCount + extraCount + diffCount
  if (total > 0) {
    const parts: string[] = []
    if (missingCount) parts.push(`${missingCount} missing path${missingCount > 1 ? 's' : ''}`)
    if (extraCount) parts.push(`${extraCount} unexpected path${extraCount > 1 ? 's' : ''}`)
    if (diffCount) parts.push(`${diffCount} mismatched content`)

    return {
      pass: false,
      message: () =>
        parts.length === 1
          ? `Found ${parts[0]}` //
          : `Found ${total} mismatches: ${parts.join(', ')}`,
      actual: actualDiff,
      expected: expectedDiff,
    }
  }

  return { pass: true, message: () => 'volumes matched' }
}

function makeEntryPreview(name: string, entry: VolumeMapEntry) {
  if (entry.type === 'empty-dir') {
    return new Directory()
  }
  if (entry.type === 'file') {
    if (isText(name, entry.data)) {
      return new TextFile(entry.data)
    }
    return new BinaryFile(entry.data)
  }
  return new Symlink(entry.target)
}

class TextFile {
  data: string
  constructor(buff: Buffer) {
    this.data = buff.toString('utf8')
  }
}

class BinaryFile {
  hash: string
  length: number
  preview: string
  constructor(buff: Buffer, trim = 32) {
    this.hash = createHash('sha1').update(buff).digest('hex')
    this.length = buff.length
    const head = buff.subarray(0, trim).toString('base64')
    const tail = buff.subarray(buff.length - trim).toString('base64')
    this.preview = `${head}...${tail}`
  }
}

class Symlink {
  constructor(public target: string) {}
}

class Directory {}
