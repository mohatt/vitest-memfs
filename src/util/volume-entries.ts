import path from 'node:path'
import isGlob from 'is-glob'
import { toRegex } from 'glob-to-regex.js'
import { isPlainObject } from './common.js'
import type { VolumePathType, VolumePathEntry } from './volume.js'

export type VolumeEntryType = Exclude<VolumePathType, 'other'> | 'any'

export interface VolumeEntryRecord {
  type?: VolumeEntryType
  count?: number
}

export interface VolumeEntryArrayRecord extends VolumeEntryRecord {
  path: string
}

export type VolumeEntries =
  | string
  | Array<string | VolumeEntryArrayRecord>
  | Record<string, VolumeEntryType | VolumeEntryRecord>

export type MatchRules = Map<string, MatchRule>

export function createMatchRules(input: VolumeEntries, prefix?: string): MatchRules {
  const basePrefix = resolvePrefix(prefix)
  const rules: MatchRules = new Map()

  function addRule(path: string, opts?: VolumeEntryRecord) {
    const entry = createMatchRule(path, basePrefix, opts)
    rules.set(entry.identifier, entry)
  }

  if (typeof input === 'string') {
    addRule(input)
  } else if (Array.isArray(input)) {
    for (const value of input) {
      if (typeof value === 'string') {
        addRule(value)
      } else if (isPlainObject(value) && typeof value.path === 'string') {
        addRule(value.path, value)
      } else {
        throw new TypeError(
          `Expected array item to be string | { path: string }, got \`${JSON.stringify(value)}\``,
        )
      }
    }
  } else if (isPlainObject(input)) {
    for (const [key, value] of Object.entries(input)) {
      if (typeof value === 'string') {
        addRule(key, { type: value })
      } else if (isPlainObject(value)) {
        addRule(key, value)
      } else {
        throw new TypeError(
          `Expected object value for key \`${key}\` to be string | plain object, got \`${JSON.stringify(value)}\``,
        )
      }
    }
  } else {
    throw new TypeError(`Expected string | array | plain object, got \`${JSON.stringify(input)}\``)
  }

  return rules
}

interface ExactMatchRule {
  kind: 'exact'
  identifier: string
  path: string
  expType: VolumeEntryType
}

interface GlobMatchRule {
  kind: 'glob'
  identifier: string
  pattern: string
  regex: RegExp
  expType: VolumeEntryType
  expCount: number
}

type MatchRule = ExactMatchRule | GlobMatchRule

function createMatchRule(rawPath: string, prefix: string, config?: VolumeEntryRecord): MatchRule {
  const { type, count } = config ?? {}
  if (type != null && !VALID_ENTRY_TYPES.includes(type)) {
    throw new TypeError(
      `Expected entry type to be \`${VALID_ENTRY_TYPES.join('`, `')}\`, got \`${type}\``,
    )
  }

  if (count != null && count <= 0) {
    throw new TypeError(`Expected entry count to be a positive integer, got \`${count}\``)
  }

  const resolved = rawPath.startsWith('/')
    ? path.posix.normalize(rawPath)
    : path.posix.join(prefix, rawPath)
  const expType = type ?? 'any'

  if (isGlobLike(rawPath)) {
    return {
      kind: 'glob',
      identifier: `glob(\`${resolved}\`)`,
      pattern: resolved,
      regex: toRegex(resolved),
      expType,
      expCount: count ?? 1,
    }
  }

  return {
    kind: 'exact',
    identifier: resolved,
    path: resolved,
    expType,
  }
}

interface DiffEntry {
  exists?: boolean
  type?: string
  count?: number
}

export function matchEntries(pathEntries: VolumePathEntry[], rules: MatchRules) {
  const actualDiff: Record<string, DiffEntry> = {}
  const expectedDiff: Record<string, DiffEntry> = {}
  const allMatches: VolumePathEntry[] = []
  let missingCount = 0
  let typeCount = 0

  rules.forEach((rule) => {
    if (rule.kind === 'exact') {
      const match = pathEntries.find(([path]) => path === rule.path)
      if (!match) {
        actualDiff[rule.identifier] = { exists: false }
        expectedDiff[rule.identifier] = { exists: true }
        missingCount++
        return
      }

      if (rule.expType !== 'any' && match[1] !== rule.expType) {
        actualDiff[rule.identifier] = { type: match[1] }
        expectedDiff[rule.identifier] = { type: rule.expType }
        typeCount++
        return
      }

      allMatches.push(match)
      return
    }

    const matches = pathEntries.filter(([path]) => rule.regex.test(path))
    if (matches.length < rule.expCount) {
      actualDiff[rule.identifier] = { count: matches.length }
      expectedDiff[rule.identifier] = { count: rule.expCount }
      missingCount++
      return
    }

    if (rule.expType !== 'any') {
      const typeMatches = matches.filter(([, type]) => type === rule.expType)
      if (typeMatches.length < rule.expCount) {
        actualDiff[rule.identifier] = { count: typeMatches.length }
        expectedDiff[rule.identifier] = { count: rule.expCount }
        typeCount++
        return
      }

      allMatches.push(...typeMatches)
      return
    }

    allMatches.push(...matches)
  })

  return {
    missingCount,
    typeCount,
    diff: { actual: actualDiff, expected: expectedDiff },
    matches: allMatches,
  }
}

function resolvePrefix(prefix?: string) {
  if (!prefix) return '/'
  const resolved = path.posix.resolve('/', prefix)
  return resolved === '/' ? '/' : resolved.replace(/\/$/, '')
}

export function isGlobLike(value: string) {
  return isGlob(value, { strict: false })
}

export function parseRegExp(value: string) {
  const match = /^\/(.{1,4096})\/([gimsuy]{0,6})$/.exec(value)
  if (match) {
    const [, expr, flags] = match
    return new RegExp(expr, flags)
  }
  return null
}

const VALID_ENTRY_TYPES: readonly VolumeEntryType[] = ['file', 'dir', 'symlink', 'any']
