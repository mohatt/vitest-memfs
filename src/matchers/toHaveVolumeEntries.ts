import { Volume } from 'memfs'
import { createMatcher } from '@/util/common.js'
import { scanVolumePaths } from '@/util/volume.js'
import { createMatchRules, matchEntries, VolumeEntries, MatchRules } from '@/util/volume-entries.js'

export interface VolumeEntriesMatcherOptions {
  /**
   * Prefix for resolving relative paths in the expected entries. Defaults to `/`.
   */
  prefix?: string
}

declare module 'vitest' {
  interface Matchers<T = any> {
    /**
     * Assert that specific paths exist in a memfs volume.
     */
    toHaveVolumeEntries(expected: VolumeEntries, options?: VolumeEntriesMatcherOptions): T
  }
}

export default createMatcher('toHaveVolumeEntries', function (received, expected, options) {
  const { utils, isNot } = this

  if (!(received instanceof Volume)) {
    return {
      pass: false,
      message: () => `Expected ${utils.printReceived(received)} to be a memfs Volume instance`,
      actual: received,
      expected: new (class Volume {})(),
    }
  }

  let rules: MatchRules
  try {
    rules = createMatchRules(expected, options?.prefix)
  } catch (error) {
    if (error instanceof Error) {
      throw new TypeError(
        `Invalid volume entries provided to ${utils.matcherHint('toHaveVolumeEntries')}. ${
          error.message
        }`,
      )
    }
    throw error
  }

  const pathEntries = scanVolumePaths(received)
  const result = matchEntries(pathEntries, rules)
  const { matches, missingCount, typeCount, diff } = result

  if (isNot) {
    // when .not, we assert that we found at least one expected entry
    const matchesCount = matches.length
    if (matchesCount > 0) {
      return {
        pass: true,
        message: () =>
          `Expected volume entries not to satisfy the expected entries, but it did ` +
          `(found ${matchesCount} match${matchesCount > 1 ? 'es' : ''})`,
        actual: Object.fromEntries(matches),
        expected: {},
      }
    }

    return {
      pass: false,
      message: () => `Volume entries did not satisfy the expected entries`,
    }
  }

  // we assert that we found all expected entries
  if (missingCount === 0 && typeCount === 0) {
    return {
      pass: true,
      message: () => 'Volume satisfied the expected entries',
    }
  }

  const counts: string[] = []
  if (missingCount) counts.push(`${missingCount} missing entr${missingCount > 1 ? 'ies' : 'y'}`)
  if (typeCount) counts.push(`${typeCount} path type mismatch${typeCount > 1 ? 'es' : ''}`)

  return {
    pass: false,
    message: () =>
      `Volume entries did not satisfy the expected entries (found ${counts.join(', ')})`,
    actual: diff.actual,
    expected: diff.expected,
  }
})
