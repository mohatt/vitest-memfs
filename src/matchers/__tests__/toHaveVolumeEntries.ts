import { describe, it, expect } from 'vitest'
import { makeTests, makeVol, VolumeInput } from '@test/util.js'
import toHaveVolumeEntries, { VolumeEntriesMatcherOptions } from '../toHaveVolumeEntries.js'

interface TestCase {
  name: string
  received: VolumeInput
  expected: any
  options?: VolumeEntriesMatcherOptions
  pass?: boolean
  not?: boolean
  throws?: boolean
}

const cases = makeTests<TestCase>([
  {
    name: 'matches an array of paths',
    received: {
      '/foo.txt': 'hello',
      '/dir': null,
    },
    expected: ['/foo.txt', '/dir'],
    pass: true,
  },
  {
    name: 'matches a single path string',
    received: {
      '/foo.txt': 'hello',
    },
    expected: '/foo.txt',
    pass: true,
  },
  {
    name: 'fails when a path is missing',
    received: {
      '/foo.txt': 'hello',
    },
    expected: ['/foo.txt', '/missing.txt'],
    pass: false,
  },
  {
    name: 'matches entry types from object input',
    received: () => {
      const v = makeVol({
        '/foo.txt': 'hello',
      })
      v.mkdirSync('/dir')
      v.writeFileSync('/dir/nested.txt', 'nested')
      v.symlinkSync('/foo.txt', '/link.txt')
      return v
    },
    expected: {
      '/foo.txt': 'file',
      '/dir': 'dir',
      '/link.txt': 'symlink',
    },
    pass: true,
  },
  {
    name: 'detects type mismatches',
    received: () => {
      const v = makeVol({
        '/foo.txt': 'hello',
      })
      v.mkdirSync('/dir')
      v.writeFileSync('/dir/nested.txt', 'nested')
      return v
    },
    expected: {
      '/foo.txt': 'dir',
      '/dir': 'file',
    },
    pass: false,
  },
  {
    name: 'supports prefix option with relative paths',
    received: {
      '/src/index.ts': 'export {}',
      '/src/utils.ts': 'export const noop = () => {}',
    },
    expected: ['index.ts', 'utils.ts'],
    options: { prefix: '/src' },
    pass: true,
  },
  {
    name: 'matches glob patterns from array strings',
    received: {
      '/src/index.ts': 'export {}',
      '/src/utils/math.ts': 'export const add = () => {}',
      '/src/styles.css': 'body {}',
    },
    expected: ['/src/**/*.ts'],
    pass: true,
  },
  {
    name: 'matches glob patterns from a string value',
    received: {
      '/src/index.ts': 'export {}',
      '/src/utils/math.ts': 'export const add = () => {}',
      '/src/styles.css': 'body {}',
    },
    expected: '/src/**/*.ts',
    pass: true,
  },
  {
    name: 'throws when array includes regular expression',
    received: {
      '/foo.txt': 'hello',
      '/bar.ts': 'export {}',
    },
    expected: [/\.txt$/],
    throws: true,
  },
  {
    name: 'matches typed glob patterns from object config',
    received: () => {
      const v = makeVol({
        '/src/index.ts': 'export {}',
      })
      v.mkdirSync('/src/components')
      v.writeFileSync('/src/components/button.tsx', 'export const Button = () => null')
      return v
    },
    expected: {
      '/src/**/*': { type: 'dir' },
    },
    pass: true,
  },
  {
    name: 'handles non-volume inputs',
    received: () => 'invalid volume' as any,
    expected: ['/foo.txt'],
    pass: false,
  },
  {
    name: 'throws for invalid path array values',
    received: {
      '/foo.txt': 'hello',
    },
    expected: ['/foo.txt', 42],
    throws: true,
  },
  {
    name: 'throws for invalid pattern objects',
    received: {
      '/foo.txt': 'hello',
    },
    expected: [{ path: 123 }],
    throws: true,
  },
  {
    name: 'throws when pattern object uses regex path',
    received: {
      '/foo.txt': 'hello',
    },
    expected: [{ path: /foo/ }],
    throws: true,
  },
  {
    name: 'throws for invalid entry type values',
    received: {
      '/foo.txt': 'hello',
    },
    expected: { '/foo.txt': 'file', '/bar': 'folder' },
    throws: true,
  },
  {
    name: 'fails when glob pattern has no matches',
    received: {
      '/foo.txt': 'hello',
    },
    expected: ['/src/**/*.ts'],
    pass: false,
  },
  {
    name: 'detects type mismatches for glob patterns',
    received: {
      '/foo.txt': 'hello',
      '/bar': null,
    },
    expected: {
      '/**/*.txt': { type: 'dir' },
    },
    pass: false,
  },
  {
    name: 'not: matches an array of paths',
    received: {
      '/foo.txt': 'hello',
    },
    expected: ['/foo.txt'],
    pass: false,
    not: true,
  },
  {
    name: 'not: fails when entries are missing',
    received: {
      '/foo.txt': 'hello',
    },
    expected: ['/missing.txt'],
    pass: true,
    not: true,
  },
])

describe('toHaveVolumeEntries()', () => {
  describe('unit', () => {
    const mockState = (isNot = false): any => ({
      utils: {
        printReceived: (received: unknown) => `received(${JSON.stringify(received)})`,
        printExpected: (expected: unknown) => `expected(${JSON.stringify(expected)})`,
        matcherHint: (matcherName: string) => `hint(${matcherName})`,
      },
      isNot,
    })

    function testRunnerUnit({ received, expected, options, pass, not, throws }: TestCase) {
      const actVol = makeVol(received)

      if (throws) {
        const matcher = toHaveVolumeEntries.bind(mockState(not))
        expect(() => matcher(actVol, expected, options)).toThrowErrorMatchingSnapshot()
        return
      }

      const invoke = () => {
        const matcher = toHaveVolumeEntries.bind(mockState(not))
        try {
          const result = matcher(actVol, expected, options) as any
          return { ...result, message: result.message() }
        } catch (error) {
          return error
        }
      }

      const result = invoke()
      if (pass != null) expect(result).toHaveProperty('pass', not ? !pass : pass)
      expect(result).toMatchSnapshot('result')
    }

    it.each(cases.normal)('$name', testRunnerUnit)
    it.only.each(cases.only)('$name', testRunnerUnit)
    it.skip.each(cases.skip)('$name', testRunnerUnit)
  })

  describe('integration', () => {
    function testRunnerInteg({ received, expected, options, pass, not, throws }: TestCase) {
      const receivedVol = makeVol(received)

      if (throws) {
        expect(() =>
          expect(receivedVol).toHaveVolumeEntries(expected, options),
        ).toThrowErrorMatchingSnapshot()
        return
      }

      if (not) {
        if (pass) {
          expect(receivedVol).not.toHaveVolumeEntries(expected, options)
        } else {
          expect(() =>
            expect(receivedVol).not.toHaveVolumeEntries(expected, options),
          ).toThrowErrorMatchingSnapshot()
        }
      } else {
        if (pass) {
          expect(receivedVol).toHaveVolumeEntries(expected, options)
        } else {
          expect(() =>
            expect(receivedVol).toHaveVolumeEntries(expected, options),
          ).toThrowErrorMatchingSnapshot()
        }
      }
    }

    it.each(cases.normal)('$name', testRunnerInteg)
    it.only.each(cases.only)('$name', testRunnerInteg)
    it.skip.each(cases.skip)('$name', testRunnerInteg)
  })
})
