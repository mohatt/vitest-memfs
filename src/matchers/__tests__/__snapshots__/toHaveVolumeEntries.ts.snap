// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toHaveVolumeEntries() > integration > 'detects type mismatches for glob patterns' 1`] = `[Error: Volume entries did not satisfy the expected entries (found 1 path type mismatch)]`;

exports[`toHaveVolumeEntries() > integration > 'detects type mismatches' 1`] = `[Error: Volume entries did not satisfy the expected entries (found 2 path type mismatches)]`;

exports[`toHaveVolumeEntries() > integration > 'fails when a path is missing' 1`] = `[Error: Volume entries did not satisfy the expected entries (found 1 missing entry)]`;

exports[`toHaveVolumeEntries() > integration > 'fails when glob pattern has no matches' 1`] = `[Error: Volume entries did not satisfy the expected entries (found 1 missing entry)]`;

exports[`toHaveVolumeEntries() > integration > 'handles non-volume inputs' 1`] = `[Error: Expected [31m"invalid volume"[39m to be a memfs Volume instance]`;

exports[`toHaveVolumeEntries() > integration > 'not: matches an array of paths' 1`] = `[Error: Expected volume entries not to satisfy the expected entries, but it did (found 1 match)]`;

exports[`toHaveVolumeEntries() > integration > 'throws for invalid entry type values' 1`] = `[TypeError: Invalid volume entries provided to [2mexpect([22m[31mreceived[39m[2m).[22mtoHaveVolumeEntries[2m([22m[32mexpected[39m[2m)[22m. Expected entry type to be \`file\`, \`dir\`, \`symlink\`, \`any\`, got \`folder\`]`;

exports[`toHaveVolumeEntries() > integration > 'throws for invalid path array values' 1`] = `[TypeError: Invalid volume entries provided to [2mexpect([22m[31mreceived[39m[2m).[22mtoHaveVolumeEntries[2m([22m[32mexpected[39m[2m)[22m. Expected array item to be string | { path: string }, got \`42\`]`;

exports[`toHaveVolumeEntries() > integration > 'throws for invalid pattern objects' 1`] = `[TypeError: Invalid volume entries provided to [2mexpect([22m[31mreceived[39m[2m).[22mtoHaveVolumeEntries[2m([22m[32mexpected[39m[2m)[22m. Expected array item to be string | { path: string }, got \`{"path":123}\`]`;

exports[`toHaveVolumeEntries() > integration > 'throws when array includes regular expression' 1`] = `[TypeError: Invalid volume entries provided to [2mexpect([22m[31mreceived[39m[2m).[22mtoHaveVolumeEntries[2m([22m[32mexpected[39m[2m)[22m. Expected array item to be string | { path: string }, got \`{}\`]`;

exports[`toHaveVolumeEntries() > integration > 'throws when pattern object uses regex path' 1`] = `[TypeError: Invalid volume entries provided to [2mexpect([22m[31mreceived[39m[2m).[22mtoHaveVolumeEntries[2m([22m[32mexpected[39m[2m)[22m. Expected array item to be string | { path: string }, got \`{"path":{}}\`]`;

exports[`toHaveVolumeEntries() > unit > 'detects type mismatches for glob patterns' > result 1`] = `
{
  "actual": {
    "glob(\`/**/*.txt\`)": {
      "count": 0,
    },
  },
  "expected": {
    "glob(\`/**/*.txt\`)": {
      "count": 1,
    },
  },
  "message": "Volume entries did not satisfy the expected entries (found 1 path type mismatch)",
  "pass": false,
}
`;

exports[`toHaveVolumeEntries() > unit > 'detects type mismatches' > result 1`] = `
{
  "actual": {
    "/dir": {
      "type": "dir",
    },
    "/foo.txt": {
      "type": "file",
    },
  },
  "expected": {
    "/dir": {
      "type": "file",
    },
    "/foo.txt": {
      "type": "dir",
    },
  },
  "message": "Volume entries did not satisfy the expected entries (found 2 path type mismatches)",
  "pass": false,
}
`;

exports[`toHaveVolumeEntries() > unit > 'fails when a path is missing' > result 1`] = `
{
  "actual": {
    "/missing.txt": {
      "exists": false,
    },
  },
  "expected": {
    "/missing.txt": {
      "exists": true,
    },
  },
  "message": "Volume entries did not satisfy the expected entries (found 1 missing entry)",
  "pass": false,
}
`;

exports[`toHaveVolumeEntries() > unit > 'fails when glob pattern has no matches' > result 1`] = `
{
  "actual": {
    "glob(\`/src/**/*.ts\`)": {
      "count": 0,
    },
  },
  "expected": {
    "glob(\`/src/**/*.ts\`)": {
      "count": 1,
    },
  },
  "message": "Volume entries did not satisfy the expected entries (found 1 missing entry)",
  "pass": false,
}
`;

exports[`toHaveVolumeEntries() > unit > 'handles non-volume inputs' > result 1`] = `
{
  "actual": "invalid volume",
  "expected": Volume {},
  "message": "Expected received("invalid volume") to be a memfs Volume instance",
  "pass": false,
}
`;

exports[`toHaveVolumeEntries() > unit > 'matches a single path string' > result 1`] = `
{
  "message": "Volume satisfied the expected entries",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'matches an array of paths' > result 1`] = `
{
  "message": "Volume satisfied the expected entries",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'matches entry types from object input' > result 1`] = `
{
  "message": "Volume satisfied the expected entries",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'matches glob patterns from a string value' > result 1`] = `
{
  "message": "Volume satisfied the expected entries",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'matches glob patterns from array strings' > result 1`] = `
{
  "message": "Volume satisfied the expected entries",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'matches typed glob patterns from object config' > result 1`] = `
{
  "message": "Volume satisfied the expected entries",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'not: fails when entries are missing' > result 1`] = `
{
  "message": "Volume entries did not satisfy the expected entries",
  "pass": false,
}
`;

exports[`toHaveVolumeEntries() > unit > 'not: matches an array of paths' > result 1`] = `
{
  "actual": {
    "/foo.txt": "file",
  },
  "expected": {},
  "message": "Expected volume entries not to satisfy the expected entries, but it did (found 1 match)",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'supports prefix option with relative paths' > result 1`] = `
{
  "message": "Volume satisfied the expected entries",
  "pass": true,
}
`;

exports[`toHaveVolumeEntries() > unit > 'throws for invalid entry type values' 1`] = `[TypeError: Invalid volume entries provided to hint(toHaveVolumeEntries). Expected entry type to be \`file\`, \`dir\`, \`symlink\`, \`any\`, got \`folder\`]`;

exports[`toHaveVolumeEntries() > unit > 'throws for invalid path array values' 1`] = `[TypeError: Invalid volume entries provided to hint(toHaveVolumeEntries). Expected array item to be string | { path: string }, got \`42\`]`;

exports[`toHaveVolumeEntries() > unit > 'throws for invalid pattern objects' 1`] = `[TypeError: Invalid volume entries provided to hint(toHaveVolumeEntries). Expected array item to be string | { path: string }, got \`{"path":123}\`]`;

exports[`toHaveVolumeEntries() > unit > 'throws when array includes regular expression' 1`] = `[TypeError: Invalid volume entries provided to hint(toHaveVolumeEntries). Expected array item to be string | { path: string }, got \`{}\`]`;

exports[`toHaveVolumeEntries() > unit > 'throws when pattern object uses regex path' 1`] = `[TypeError: Invalid volume entries provided to hint(toHaveVolumeEntries). Expected array item to be string | { path: string }, got \`{"path":{}}\`]`;
